0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,82 @@
+# Pre-Shared Key
+
+The Pre-Shared Key (PSK) protocol is an end-to-end transport protocol, used by the sender and receiver of an ILP payment to decide on a condition and fulfillment for a payment.
+
+As the name suggests, PSK relies on a pre-shared secret key that is shared beforehand. An application-layer protocol like SPSP could pass this pre-shared secret key over HTTPS. Within one node process, the pre-shared key could be passed via a simple variable.
+
+An advantage of PSK is that the pre-shared secret key only needs to be shared once. So long as both parties possess this key and are listening for transfers, they can send payments between one another.
+
+A disadvantage of PSK is that it is repudiable. Knowledge of a payment's fulfillment cannot be used as cryptographic proof that the receiver completed the payment, because both sender and receiver know how to generate the fulfillment with their pre-shared secret key.
+
+## Flow
+
+0. The pre-shared secret key is shared out of band.
+0. The sender creates an inner (PRIVATE) PSK request, with information on the expiry and any additional data enclosed within.
+0. The sender creates a random Key-Id value, and calculates an HMAC of the Key-Id using the pre-shared secret key as the key. This HMAC will be referred to as the payment key.
+0. The sender encrypts the inner request with the payment key.
+0. The sender creates an outer (PUBLIC) header, and attaches the Key-Id and HMAC function they used as headers. The encrypted inner request is attached as binary data.
+0. The sender quotes to the receiver and sends a transfer with the outer header in the data field of the ILP packet. The fulfillment of this payment is the HMAC of the ILP packet using the payment key as the key.
+0. The receiver gets notified of in incoming prepared transfer with an ILP packet.
+0. The receiver generates the payment key from the outer request contained in the data field of the ILP packet, and uses it to decrypt the inner request.
+0. _(Optional)_ The receiver calls a backend to review the payment.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key as the key.
+
+## Data Format
+
+The PSK details are divided into two request objects in the style of [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). The data of the outer request is the encrypted inner request.
```

Comment:
If we're using "HTTP-style" headers it would probably be best to use their format exactly